text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,592 @@
+- `CARGO_BUILD_RESOLVER_PRECEDENCE=rust-version` assumes maximal resolution as generally minimal resolution will pick packages with compatible rust-versions as rust-version tends to (but doesn't always) increase over time.
+  - `cargo add` selecting rust-version-compatible minimum bounds helps
+  - This bypasses a lot of complexity either from exploding the number of states we support or giving users control over the fallback by making the field an array of strategies.
+
+## Add `workspace.rust-version`
+
+Instead of using the lowest MSRV among workspace members, we could add `workspace.rust-version`.
+
+This opens its own set of questions
+- Do packages implicitly inherit this?
+- What are the semantics if its unset?
+- Would it be confusing to have this be set in mixed-MSRV workspaces?  Would blocking it be incompatible with the semantics when unset?
+- In mixed-MSRV workspaces, does it need to be the highest or lowest MSRV of your packages?
+  - For the resolver, it would need to be the lowest but there might be other use cases where it needs to be the highest
+
+The proposed solution does not block us from later going down this road but
+allows us to move forward without having to figure out all of these details.
+
+## Reporting
+
+Alternative to or in addition to the `cargo update` output to report when things are held back (both MSRV and semver),
+we can run the resolver twice on the original input, once for MSRV and once without.
+We then do a depth-first diff of the trees, stopping and reporting on the first different node.
+This would let us report on any command that changes the way the tree is resolved.
+We'd likely want to limit the output to only the sub-tree that changed.
+
+We could either always do the second resolve or only do the second resolve if the resolver changed anything,
+whichever is faster.
+
+Its unknown whether making the inputs available for multiple resolves would have a performance impact.
+
+While a no-change resolve is fast, if this negatively impacts it enough, we
+could explore hashing the resolve inputs and storing that in the lockfile,
+allowing us to detect if the inputs have changed and only resolving then.
+
+## Make this opt-in
+
+As proposed, CI that tries to verify against the latest dependencies will no longer do so.
+Instead, they'll have to make a change to their CI, like setting `CARGO_BUILD_RESOLVER_PRECEDENCE=maximum`.
+
+If we consider this a major incompatibility, then it needs to be opted into.
+As `cargo fix` can't migrate a user's CI,
+this would be out of scope for migrating to this with a new Edition.
+
+I would argue that the number of maintainers verifying latest dependencies is
+relatively low and they are more likely to be "in the know",
+making them less likely to be negatively affected by this.
+Therefore, I propose we consider this a minor incompatibility
```

Comment:
This section doesn't talk about the option of keeping this opt-in *indefinitely*. You talk about making it opt-in *temporarily* and eventually move to changing the default, and I agree that that would be worse. But could you please add some discussion of the alternative of making the MSRV-compatible resolver available as a non-default option for projects with old MSRVs to opt into?

Reply:
@epage For this, could you also add the note that we could still make `cargo add` check MSRV and avoid setting a version constraint that would be incompatible with `rust-version`? You noted that you wanted to have support in `cargo add`, and it seems worth noting that so that the defaults don't feel all-or-nothing.